### ISSUE TITLE
Ability to download old releases

### ISSUE THREAD

**Issue by mfornasa** [docker/for-mac#1120]
I am aware of https://github.com/docker/for-mac/issues/444, but that issue has been closed without explanation.

It should be possible to download older versions. For example, version 1.13.0-rc5 has a high priority  bug (https://github.com/docker/docker/issues/29950) that breaks all our caching. rc5 has been auto installed, and I am unable to find the package to revert to rc4 while waiting for a new stable release with the fix.

Please advise.

**Comment by justincormack** [comment]
@mfornasa 1.13.0-rc5 is a beta channel release, so unfortunately there may be breakages. There will be a new beta release this week. If you want a stable release, we recommend installing from the stable channel.

**Comment by mfornasa** [comment]
I accept the possibility of a breakage on beta, no issues with that.

I would like to be able to **manually** download an older package. This would also help in debugging issues with the beta and provide you better feedback.

**Comment by mfornasa** [comment]
@justincormack I have been able to find the link for version rc4 using the forum: https://download.docker.com/mac/beta/1.13.0.14840/Docker.dmg. I still believe that it would be useful to have a list of those links somewhere.

**Comment by cloakedcode** [comment]
There are two ways to downgrade:

## 1. Recreate the URL
If you happen to know the build number (e.g. 14937) -- not the git commit you see when running `docker --version` or `docker version` -- you can form the URL:

```shell
https://download.docker.com/mac/{{ beta | stable}}/{{ version }}.{{ build number }}/Docker.dmg
# example:
https://download.docker.com/mac/stable/1.12.6.14937/Docker.dmg
```

(Note: the build number shows up in "About Docker" from the menulet menu.)

![screen shot 2017-01-23 at 2 12 56 pm](https://cloud.githubusercontent.com/assets/177435/22225134/260a59a4-e176-11e6-8fdf-0ecc556f3898.png)

Of course, you'd need to know the build number, which is hard to do if you've already upgraded.

## 2. Pull it from Trash
After upgrading, the old `Docker.app` gets put in the Trash (e.g. `Docker (14937)`). Uninstall the current version, Preferences... -> Uninstall / Reset -> Uninstall; delete `/Applications/Docker`; copy the old version from the Trash into `/Applications`; rename to `Docker`; and launch `Docker` to begin the install process.

**Comment by hyperknot** [comment]
@cloakedcode's link for 1.12.6 is the _only_ code on the internet for a reliable 1.12 version.

This is a joke. Pushing an unstable, buggy version and then just not caring about keeping an archive download list.

1.12.6 is actually totally missing from the Release Notes, so no way to figure out the build number ourselves.

**Comment by friism** [comment]
@hyperknot there was never a stable release that included Docker Engine 1.12.6. 1.12.6 and 1.13.0 were released almost at the same time so Docker for Mac stable went straight to 1.13.0.

Are you experiencing stability problems with Docker 1.13?

**Comment by mrhanlon** [comment]
@friism I'll speak for myself at least, but I found 1.13 to be unusable on my corporate network. It would not play nice __at all__ with VPN/Proxy/DNS, where the 1.12.x releases were fine.

I will say that this is somewhat beside the point here. I don't think that it is unexpected that at least __someone__ is going to have a problem with an update release affecting their workflow. But when that does happen, the most common thing to do is go download the last known working version. But since Docker for Mac doesn't list any archived releases that is impossible, unless you magic the URL out of the Github issues.

**Comment by mfornasa** [comment]
To be clear: I am not suggesting an automatic downgrade mechanism, or nothing fancy. A simple **list**  of download links for previous version would be really valuable. The lack of it has been a nuisance for me and my company twice in the last couple of months.

**Comment by hyperknot** [comment]
@mrhanlon https://download.docker.com/mac/stable/1.12.6.14937/Docker.dmg is a stable release with 1.12.6 and works reliably.

Yes, 1.13 is full of bugs, I simply could delete exited containers on it and so on. It was like that with 1.12.(0,1,2) and I believe it'll be like this in the future.

**Comment by friism** [comment]
@mrhanlon @hyperknot if you haven't already, can you run diagnostics when this is failing and open a separate issue so that we can fix the problem?

We're continuously fixing bugs (including security-related problems), and those fixes are released in new versions of Docker for Mac. We don't support or patch older versions - we only support the latest version on the stable and beta channels respectively.

Because we want users to be running software that's patched and secure, we don't publicly document how to download and install old versions.

If you find bugs in Docker for Mac and Windows (either stable or beta), please run the in-app diagnostics and open an issue so that we can address the problem.

**Comment by mrhanlon** [comment]
@friism Thanks. There are/were several tickets already open related to the DNS/Proxy issues I was experiencing. But the fact remained that I was pretty much 100% blocked by this problem and needed a way to roll back pronto.

**Comment by jelder** [comment]
@friism There will always be reported, unfixed bugs. If you don't want to "support" old versions that's fine. I know you have limited resources.

However, it is inconsistent to allow users to selectively turn off automatic updates while not allowing downloads of previous versions. 

![image](https://cloud.githubusercontent.com/assets/38099/22671340/d8c50bd6-ec9b-11e6-9ae4-4c259ac878fc.png)

Would it be so bad if you just enabled `S3:ListBucket` on download.docker.com?

**Comment by phbcanada** [comment]
@mrhanlon - I was able to work around the network bug (on Mac) by quitting Docker - move the old version from Trash to Applications and then run the old version. Probably only a temporary fix but I was able to get my images updated. I was lucky to still have the old version there.

**Comment by shawkinsl** [comment]
It's totally wild that this issue is still unresolved. Boy howdy for once am I glad I never de-clutter my downloads folder.

**Comment by if-kyle** [comment]
+1 as someone who helps maintain d4m-nfs, it makes it very hard to know when users upgrade, to know if its a breakage in our code or something that has changed in docker for mac.

**Comment by franklinyu** [comment]
@cloakedcode's trick no longer works with latest version scheme starting with 17: https://download.docker.com/mac/edge/17.05.0.17656/Docker.dmg doesn't work.

**Comment by zbruhnke** [comment]
in case this helps anyone else I put an older version of the docker edge beta in an S3 bucket here since i had one in trash

This one in particular is reflected in the filename

17.05.0-ce-rc1-mac8-build16582

If it helps you're welcome to it - I'll leave this up for a while

https://s3-us-west-2.amazonaws.com/docker-edge/Docker-edge-17.05.0-ce-rc1-mac8-build16582-git-sha-73d01bb48e.dmg

**Comment by subfuzion** [comment]
@zbruhnke *DISCLOSURE: I'm not a lawyer and I'm not authorized to speak for Docker, just offering my own personal interpretation of the license agreement*. As helpful as you are trying to be, I would not personally recommend doing what you're doing as it is quite possibly in violation of Docker's End User License Agreement.

The relevant excerpts from that agreement would seem to be (emphasis mine):

> 1.1.	Licensed Software.  Subject to Customer’s compliance with the terms and conditions of this Agreement, Docker hereby grants Customer a perpetual, revocable, non-exclusive, non-transferable, non-sub-licensable license to *download, install and use the Licensed Software on a single computer*, solely in object code format, and solely for  Customer’s internal business purposes.

> 2.	Restricted Activities.  Customer shall not, and shall not encourage any third party to:  ... (c) *distribute*, license, sublicense, lease, rent, loan, *or otherwise transfer the Licensed Software to any third party;* ...

I have run into occasionally run into regressions with the latest edge release and sometimes it's convenient to revert back to the previous without switching to stable. Probably the best thing is to just restore the version that was moved into the trash when you updated.

If you remember to check and save the build number, you can also use that to fetch another copy of the installer:

https://download.docker.com/mac/edge/{{build}}/Docker.dmg

For example:

Version 17.05.0-ce-mac11 (**17656**)
Channel: edge
ffbc5f5871
https://download.docker.com/mac/edge/17656/Docker.dmg

Again, the best option is to just restore the previous version from your trash folder since as far as I know the specific URL pattern isn't supported and might change in the future.

**Comment by zbruhnke** [comment]
@subfuzion Thanks for the heads up - in this case I'm not sure if anyone was unhappy about it - but I assumed someone else may be using an older edge version to do something like connect to their swarms and deploy and since that hindered my ability to do so I assumed it may help someone else.

If it is in violation and someone from Docker, Inc. was to let me know I would happily remove it - I am not maintaining an illegal mirror of all downloads or anything like that - just trying to help someone else use some OSS they may need to have something critical working.

Nothing I was doing here was nefarious to be certain and I was never planning on leaving it up indefinitely

**Comment by jelder** [comment]
The continued activity on this issue with no feedback from Docker is pretty disappointing. It shouldn't be the case where people are forced to do dubiously legal things just to get working versions of this thing.

Is Docker for Mac abandoned?

**Comment by friism** [comment]
@jelder Docker for Mac is not abandoned, and I provided feedback on this issue [previously](https://github.com/docker/for-mac/issues/1120#issuecomment-277774579).

We encourage users to download Docker software from Docker.

**Comment by franklinyu** [comment]
I created [a Gist](https://gist.github.com/franklinyu/5e0bb9d6c0d873f33c78415dd2ea4138) to manually record the links myself. Only latest few versions are recorded. (No one wants to revert to 1.13.0 these days, right?)

**Comment by buchanae** [comment]
I'm in the middle of a work day here and after upgrading to 17.06 (accidentally, because of the intrusive pop-up while I was typing), my computer first crashed, and then my test suites started crashing frequently. I *need* to downgrade, and I don't have time to play debugger for docker.

**Comment by djs55** [comment]
@buchanae in case it's useful, here's a link to [17.03.1-ce-mac5 (16048)](https://download.docker.com/mac/stable/16048/Docker.dmg)

**Comment by friism** [comment]
@buchanae before downgrading can you please try generating a diagnostic dump so that we can debug the 17.06 problem you're encountering?

**Comment by buchanae** [comment]
@djs55 Thanks. I ended up digging it out of the Trash bin.

@friism Yes, I clicked the "diagnose & upload" button. Is that good enough? Unfortunately, all the lights were green, but there might be details in there. I'm happy to help debug when I am free.

**Comment by friism** [comment]
@buchanae Ideally we'd need the diagnostic ID that was generated, but no problem if you just want to get on with your day.

**Comment by christhomas** [comment]
it would appear that its such an easy problem to fix, running software like this is always prone to something breaking, but providing a quick and easy menu option to select the version you want to install and use would solve the problem instantly whilst letting people try out new versions with their software and quickly regressing back to a previous version if there is a hiccup.

its like a combo drop down box and a list of download links and execute the install process, I don't see such a big deal here, we are all developers, so this seems like childs play.....

so whats the hold up?

**Comment by lnewportTX** [comment]
I'm glad to have found this.  I'm receiving the following error " unknown shorthand flag: 'e' in -e"  after upgrading to to 17.06.0 and I'm not a savvy enough user to fix this.    I know where the -e is coming from but the script works with previous docker versions.

**Comment by friism** [comment]
@christhomas the hold-up is that Docker Community Edition 17.03 is no longer maintained or updated. It may thus have security issues or bugs that are fixed in Docker CE 17.06. More on the release cycle here: https://blog.docker.com/2017/03/docker-enterprise-edition/

@lnewportTX the `-e` flag was deprecated in Docker 1.11 and removed in Docker 17.06: https://docs.docker.com/engine/deprecated/#-e-and---email-flags-on-docker-login

What script are you running that's showing this error? You should work with the script author to update.

**Comment by christhomas** [comment]
ok great, so this feature is going into 17.06 I guess or where? is there a specific bug ticket to add this, or should we add one, what do you need from us in order to make a request for this feature, I think its about half a days work to add a checkbox and a download archive isnt it?

this conversation is from january and yet in an afternoon it could have been done already...thats kind of my point, instead of asking continuously for 7 months about it, put it in the bug list, then we can see when it might get released, then we dont have to talk about it anymore, apart from asking the status, or maybe to upgrade its urgency because you are constantly breaking releases and not giving a rollback mechanism, its highly infuriating.

**Comment by hyperknot** [comment]
Look, without any negative tone it is clear that Docker is now trying to sell stability as a monetisation strategy. That means you are supposed to be always running on bleeding edge release and provide feedback so they can take polish it and make a stable release, which they can sell for paying customers. When the stable release is done you are already testing the next free release. 

Where this gets really said is how they actively work against making a download archive of old releases. It really is quite an aggressive strategy on their behalf.

So we are left here trying to figure out obscure URL locations and trying to mirror those releases ourselves, questioning even if it's legal to do so.

All that, while their landing page has a whole dedicated section to open source in the footer pointing to a page dedicated to open source:

> "Our open source projects serve as a basis for our free and open Docker platform. In the last two years, over 450 million people worldwide have downloaded Docker’s open source software."

**Comment by christhomas** [comment]
I dont mind them breaking stuff if its easy for me to downgrade back to a previous version where my stuff was working before, I am a programmer, we pretty much all are, right?

So ok, add a menu, select a version, click "install" and bingo, problem solved, what about a "lock" checkbox too maybe? 

Then they can go ahead and break whatever they like and when I am in the period of time allocated to me to dedicate to keeping things fresh, thats what I will do and I will upgrade my code, its not a problem, its absolutely ok, just give me an easy way to fix the problem, then I am happy and with half a days work it could be done, so I dont know what this 7 month conversation has been about really, seems ridiculous to me.

**Comment by isaacuribe** [comment]
Man, this annoying, so much that I will bother to write in here.

No software is free of bugs, therefore, any given combination of software programs will be riddled with bugs, if you take into account network, cosmic rays, libraries, and magnetism, at any given time you are working on top of hundreds, maybe over thousands of things that can go wrong, "Docker for Mac" is only one of them.

The cycle of my life as an IT engineer, for the last ~20 years:

1. Invest blood, sweat, and tears until you find the combination of software versions that perform good enough, securely enough, and stay out of your way as a foundation for whatever you are after
2. Do your work, deal with even more problems, fix them
3. Ship
4. Hopefully, make some money
5. GOTO 1

Do I have the time to report every bug and upgrade every single piece of software every day? Of course not!, I wouldn't be able to get past step 1.

Pushing you to only run the latest release is unreasonable, just as me, you have limited bandwidth, it makes sense NOT to support old versions, do ignore people with unreasonable requests, especially non-paying customers, don't invest time back-porting patches, I understand you are trying to find a way to make money, we all are, it is reasonable to get the most bang for your time.

But at least accept that "stable" != "latest", "stable" is what I say is stable for me so I can ship my work, not what you label "stable" after the limited testing you can do (I am not being pretentious, you are doing great work, don't take it as an offense, it is simple logic: (your test coverage) < (the sum of use cases that we, the users in the rest of the world, are coming up with)

So yes, the latest and shiniest version of Docker for Mac crashed for the 9th time in a period of 2HRS after I upgraded (my mistake), I don't care about reporting the problem, I don't have the time, the latest version on my datacenters is working fine, but I had to spend another hour of my limited time looking for a way to downgrade, until I finally found the installer on another laptop's trash can.

And here I am, waiting until all my builds finish while I write this, whatever is broken keeps corrupting my "Docker.qcow2" file (yes, I tried deleting and starting from scratch), for my definition of "stable", the old version is "stable" and the new version is unreliable, it just cost me 3HRS of my life / $750 of unbillable hours.

For the record, I do appreciate your work, I am a paying customer, and I do report bugs when I have the time and think it is a good case, but please understand you are just making me spend more time than it is necessary by not providing a simple list of links to a handful of previous releases.

Just drop them on a cheap hosting behind an ugly, Apache-generated auto-index HTML file (or S3, or whatever CDN option is free for you) and post the link here ;-)

**Comment by gigafied** [comment]
Honestly, just keep a README in the repo with links to the old versions.

17.06.0-ce broke ADD behavior with urls, it's been acknowledged as a bug and promised in 17.06.1, that was over a month ago. If you're only going to publish the latest "stable" versions, you can't also have a practice of not fixing breaking bugs for months.

**Comment by jsok** [comment]
You can scrape the S3 bucket for releases, a quick bash script like this:

```bash
channel=stable
for build in $(seq 19125 1800); do
  if curl -X HEAD -fsSL -I https://download.docker.com/mac/$channel/$build/Docker.dmg >/dev/null 2>/dev/null; then
    echo "Found build: $build"
    curl -X HEAD -fsSL -I https://download.docker.com/mac/$channel/$build/Docker.dmg
  fi
done
```

The numbers in the `seq` are build numbers, you can use the update XML file(s) to figure out what the latest build is, see:

 * https://download.docker.com/mac/stable/appcast.xml
 * https://download.docker.com/mac/edge/appcast.xml

Example metadata in the response header:

```
HTTP/1.1 200 OK
Content-Type: binary/octet-stream
Content-Length: 129940393
Date: Wed, 13 Sep 2017 06:12:23 GMT
x-amz-meta-humanversion: 17.06.1-ce-mac24 (19058)
x-amz-meta-sha1: 46ba614c96da572a08a3cd29767ac7bdadf23caa
x-amz-meta-channel: stable
x-amz-meta-version: 17.06.1-ce-mac24
x-amz-meta-checksum: cff09dfbc0cdbe99e513ad58e517e7420471785b34554002d27c85f467c0a13a
x-amz-meta-arch: mac
x-amz-meta-build: 19058
Last-Modified: Fri, 01 Sep 2017 13:05:00 GMT
x-amz-version-id: _WKiFmeMOjr2v5eE8jV847FzcWUFV.3p
ETag: "6acdb8e1731cc3cd8e7dd76326679103-25"
Server: AmazonS3
Age: 167
X-Cache: Hit from cloudfront
Via: 1.1 4f2b51b0906eb4177f90fe010732e8a3.cloudfront.net (CloudFront)
X-Amz-Cf-Id: SX8PsTGu_J1dM0S2kCC7nbkT9QMbePfkGRIYuRvrKsgBIc0cChY9ng==
Proxy-Connection: Keep-Alive
Connection: Keep-Alive
```

We should scrape and publish the old release links to a github repo/gist.

**Comment by mastermindg** [comment]
@jsok Your script only catches new builds. How can I download 1.13 stable?

**Comment by jsok** [comment]
@mastermindg no idea mate, you can do some more poking around the S3 bucket(s) and might get lucky.

**Comment by franklinyu** [comment]
@mastermindg Just curious, why would you like 1.13, given "1.13.0 is full of bugs" from @hyperknot?

**Comment by mastermindg** [comment]
Why would you assume I wanted 1.13.0?

**Comment by franklinyu** [comment]
@mastermindg Sorry, I meant 1.13. There are two stable series, 17.03 and 17.06, released after 1.13. Most users following this issue seem to expect only "last working version", and that's why I was asking.

**Comment by mastermindg** [comment]
I've been running 1.13.1 for months now without any issues. I run this
version on my production systems because Kubernetes is bound to it. I only
test containers on my laptop but it would be nice to go back as there are
no new native features that I'm using and compatibility is important.

**Comment by franklinyu** [comment]
@mastermindg I assume that all release in 1.13 series have reached end of life, unless you are using the [Docker CSE (Commercially Supported Engine)](https://docs.docker.com/v1.13/cs-engine/1.13/). For better support for old versions (since most enterprise would prefer stability over functionality), I guess the Docker team would recommend CSE (now called [Docker EE](https://www.docker.com/enterprise-edition)). I never mean that Docker should not support downloading 1.13, but it is even less likely than downloading "last working version".

**Comment by u007** [comment]
still an on going issue?
it just broke on docker 17.09.0-ce-mac32 (19506) on both mac sierra and high sierra

**Comment by SamFromDaUk** [comment]
I ran into a similar issue where the I upgraded from `17.03.1-ce` to `Docker 17.09.0-ce`. This caused my docker configuration to break with an error of `nc: bad address 'database-product'`. Im pretty sure this will be an issue with our configuration, but we cant always drop tools and start debugging Docker immediately.

To temporarily fix this I spent around an hour trying to find what my previous version of Docker was, and then downgrade Docker back to this version. Eventually I ended up on this thread where @cloakedcode had the solution for me (check your trash for the previous version of Docker).

I know my team would definitely appreciate better signage in 2 areas. 

1) How to find out your previous version of Docker after an upgrade
2) How to downgrade to the previous version after a failed upgrade (for whatever reason)

**Comment by douglascamata** [comment]
Why is this not moving any forward? I was already able to find some hidden links here: https://download.docker.com/mac/static/stable/x86_64/. We just need a webpage in front of it or a link for it somewhere. Quite simple, requires no discussion at all and it's super annoying when you need to rollback a version (doesn't matter why).

**Comment by flaccid** [comment]
@douglascamata my guess is that Docker Inc. is becoming another Atlassian i.e. doesn't listen to customers. The most demanded is ignored.

**Comment by if-kyle** [comment]
Got burned by this today, latest version of docker for mac fails to stop or rm all my containers.  Found an older version from https://gist.github.com/franklinyu/5e0bb9d6c0d873f33c78415dd2ea4138 and that fixed it.

**Comment by mdgreenwald** [comment]
This really shouldn't be a problem. Just post a list of stable releases somewhere!

**Comment by cloudbow** [comment]
The docker mac networking stopped working with "lib curses protocol error" today and I was unable to find an older release.

**Comment by jsok** [comment]
I've made some progress on automating the scraping of download URLs. It needs to be hooked up to a Travis CI job and run daily, but it's publishing to GitHub Pages and scraped data is a JSON file for each channel that Jekyll can parse and group by quite nicely.

You can find the page here: https://jsok.github.io/docker-for-mac-versions/

**Comment by christhomas** [comment]
wow....the fact that some random guy from the internet has to fix an issue which the entire docker team apparently struggles with for months on end and the solution is so bloody simple, kudos to @jsok for showing the docker team up and hoping that they finally get their shoddy act together, well done!

**Comment by friism** [comment]
@christhomas as mentioned previously, not publishing previous releases is a [conscious decision](https://github.com/docker/for-mac/issues/1120#issuecomment-277774579).

> We're continuously fixing bugs (including security-related problems), and those fixes are released in new versions of Docker for Mac. We don't support or patch older versions - we only support the latest version on the stable and beta channels respectively.

> Because we want users to be running software that's patched and secure, we don't publicly document how to download and install old versions.

**Comment by christhomas** [comment]
Yeah, you keep breaking stuff, then leaving us high and dry whilst you fix those bugs without a clear way to go back, you can colour this how you want, but you basically are doing a piss poor job that some guy from the internet has to fix your problems cause you consciously are doing a bad job.

If you want to move fast and break things, thats fine, but whilst your stuff is broken, you should provide a way for me to continue working, otherwise, you eventually will lose an edge user who comes here and drops off reports of things being broken, because the time taken to fix the issue and migrate back to the previous version is just not worth the hassle.

it's a bad decision, one which has been told to you time and time again, yet you are too arrogant to admit it.

the solution is easy, simple, and was provided by us, because YOU ARE NOT LISTENING TO OUR NEEDS.

contemplate that for a moment before you reply back "we understand the criticism, however, blah blah blah".....nobody cares what your reasoning is, we are professionals, losing time and hair because of this issue and the solution is a five minute effort on your part and the solution is ALWAYS THE SAME, to downgrade manually, jfc, can you be anymore obtuse?  At this point, you are just not being professional.  You are taking it personally and not doing the right thing cause of your inflated egos

**Comment by christhomas** [comment]
And before you come back with another pithy yet tone deaf reply, consider this, nobody expects to run outdated, unpatched software, thats not the issue here.

The issue is that today you might release a version which breaks my setup, tomorrow it will be fixed, but today its broken, my computer will offer to install it, I will agree cause I want the leading edge and my setup is broken.

For today, so I can not lose time and effort, I need to quickly downgrade, then when you bring out another version, I can upgrade to it to see if that has fixed my issue.

It has nothing to do with wanting to run outdated, or unpatched software, it has everything to do with wanting to continue doing my work and not wasting time on this.

Is this really so difficult for you guys to understand, since it has been told to you for months and yet, here we are, some random guy fixes it for you and makes everybodies lives easier, whilst you sit in your ivory tower, breaking stuff and not taking responsibility for it.

**Comment by SamFromDaUk** [comment]
I wholeheartedly disagree with your decision on this @friism. I'd like you to listen to the community on this and do what the vast majority of open source software does - offer older release versions!

If you think this may lead to a higher percentage of users running outdated software, the put a big notice at the top of the "download prior releases" page saying that they should upgrade to X version as soon as possible to fix security issues.

To be honest I would love it if it was built into Docker for Mac. But having a download page seems like a good middle ground for now.

**Comment by iolloyd** [comment]
@friism Nobody is asking you to patch older software, just provide an easy way to locate previous versions. Is it so bad to provide this? From reading this thread, it seems that there would be a very grateful audience, who although acknowledging your point of view regarding fixing bugs for new versions only, would be able to save hours of valuable time manually rolling back.

**Comment by jsok** [comment]
I've opened https://github.com/docker/docker.github.io/pull/5044 which begins the initiative of adding the download links to their respective release notes.

**Comment by blasterpal** [comment]
#latetotheparty  I wanted to downgrade to an earlier version to address slow volume mounts that suddenly started. So I stumbled onto another method to obtain URLs for older Docker for Mac downloads: **Use Homebrew repo to lookup the builds and download.**
For example: https://github.com/caskroom/homebrew-cask/commit/68eaa0382a8d63693798f8302f3f4b65b1d8571a#diff-a0d0875b383dcea25030e97ef8b38e31

Homebrew is widely maintained project and I'd suspect this is fairly reliable means.

**Comment by franklinyu** [comment]
@iolloyd I'm pretty pessimistic about what the team would do. This issue will end up being a discussion thread about how the community (especially @jsok) hack it out.

**Comment by friism** [comment]
Thanks for all your feedback on whether to make old Docker for Mac and Windows download links available.

We're taken your criticism to heart. Acknowledging that new stable builds occasionally break users we'll include links to old installers in the release notes. We've merged @jsok's [PR](https://github.com/docker/docker.github.io/pull/5044/files) and will also include download links in release notes for future releases.

I'll close this issue - please let us know if this doesn't resolve the problem.

**Comment by iolloyd** [comment]
@friism much appreciated, thank you.

**Comment by franklinyu** [comment]
It would be much appreciated if old releases of Edge channel are also provided. This is actually good to Docker, since it encourage us to use Edge and report bugs.

**Comment by hyperknot** [comment]
@jsok really nice work! But how did you get the matching full name (17.09.0-ce-mac35) <> build number? By hand? The Github script you have doesn't seem to have this automatically and I've also been struggling to find a source for matching build numbers to full names.
My best ideas so far had been homebrew repo and Chocolately index:
https://chocolatey.org/packages/docker-for-windows

**Comment by franklinyu** [comment]
@hyperknot The HTTP response header has `x-amz-meta-version: 17.09.0-ce-mac34 (19605)`.

**Comment by hyperknot** [comment]
@franklinyu those were added only recently, for example 17.03 (https://download.docker.com/win/stable/10743/Docker%20for%20Windows%20Installer.exe) didn't have them.

**Comment by franklinyu** [comment]
@hyperknot Oh you are talking about Windows?

**Comment by hyperknot** [comment]
@franklinyu both. The S3 headers started appearing at 17.06 and were not present before 17.03:
https://download.docker.com/mac/stable/15583/Docker.dmg

**Comment by franklinyu** [comment]
Then it makes sense; that's why @jsok only got versions starting from 17.06.

**Comment by hyperknot** [comment]
But the PR to add Download links to Release Notes actually contained them all. I guess it must have been manual work then.
https://docs.docker.com/docker-for-mac/release-notes

**Comment by gcstang** [comment]
Does anyone have 17.12.0 rc1 ?

**Comment by yuliyanm** [comment]
Yesterday I updated to 17.12.0-ce-mac46 (21698) and now I can not push and pull from privete usecure registries any more, and I have two deadline pressing on my neck. Why does Docker not provide an offcial way to downgrade!!!!!!!!!

**Comment by hyperknot** [comment]
@yuliyanm they do, just download from https://docs.docker.com/docker-for-mac/release-notes/#stable-release-notes

**Comment by gcstang** [comment]
@yuliyanm 
This is the 17.11.0-ce-mac40 (20561) version.
https://download.docker.com/mac/edge/20561/Docker.dmg

**Comment by IvanLinos** [comment]
Hi,
There still no way to find the build number for edge releases and so, downgrade it to a given version.
This last 18.02.0-ce-mac53 release totally broke my environment, but 18.02.0-ce-rc1-mac50 worked like charm! Why dont you provide download links for edge releases???
We would need at least a way to identify the build number so we can build the download link as:
https://download.docker.com/mac/edge/BUILDNUMBER/Docker.dmg

**Comment by akimd** [comment]
I don't understand well all the messages here.  As @hyperknot already reported, you can still download the previous releases from the release pages:
- https://docs.docker.com/docker-for-mac/release-notes/
- https://docs.docker.com/docker-for-mac/edge-release-notes/
- https://docs.docker.com/docker-for-windows/release-notes/
- https://docs.docker.com/docker-for-windows/edge-release-notes/
